jobsdb/jobsdb.go

```diff
@@ -2377,57 +2399,73 @@ func (jd *HandleT) addNewDSLoop(ctx context.Context) {
 			return
 		case <-jd.TriggerAddNewDS():
 		}
-
-		// Adding a new DS only creates a new DS & updates the cache. It doesn't move any data so we only take the list lock.
 		var dsListLock lock.LockToken
 		var releaseDsListLock chan<- lock.LockToken
-		// start a transaction
-		err := jd.WithTx(func(tx *Tx) error {
-			return jd.withDistributedSharedLock(context.TODO(), tx, "schema_migrate", func() error { // cannot run while schema migration is running
-				return jd.withDistributedLock(context.TODO(), tx, "add_ds", func() error { // only one add_ds can run at a time
-					var err error
-					// We acquire the list lock only after we have acquired the advisory lock.
-					// We will release the list lock after the transaction ends, that's why we need to use an async lock
-					dsListLock, releaseDsListLock, err = jd.dsListLock.AsyncLockWithCtx(ctx)
-					if err != nil {
-						return err
-					}
-					// refresh ds list
-					var dsList []dataSetT
-					var nextDSIdx string
-					// make sure we are operating on the latest version of the list
-					dsList = getDSList(jd, tx, jd.tablePrefix)
-					latestDS := dsList[len(dsList)-1]
-					full, err := jd.checkIfFullDSInTx(tx, latestDS)
-					if err != nil {
-						return fmt.Errorf("error while checking if DS is full: %w", err)
-					}
-					// checkIfFullDS is true for last DS in the list
-					if full {
-						if _, err = tx.Exec(fmt.Sprintf(`LOCK TABLE %q IN EXCLUSIVE MODE;`, latestDS.JobTable)); err != nil {
-							return fmt.Errorf("error locking table %s: %w", latestDS.JobTable, err)
+		addNewDS := func() error {
+			defer func() {
+				releaseDsListLock <- dsListLock
+			}()
```

Comment:
both `dsListLock` and `releaseDsListLock` are being assigned at a later stage and might not be in case of an error. Thus they might be nil at the time of the defer statement is executed

Reply:
added nil check in defer.